`AssertionError: kind(arg) == K"Identifier" || is_operator(arg)` in broadcasting over `Diagonal`s

```julia
julia> using LinearAlgebra

julia> @descend_code_warntype (() -> cos.(Diagonal(1:4)))()
(::var"#3#4")() @ Main REPL[2]:1
┌ Warning: couldn't retrieve source of (::var"#3#4")() @ Main REPL[2]:1
└ @ TypedSyntax ~/.julia/packages/TypedSyntax/uinhI/src/node.jl:31
Variables
  #self#::Core.Const(var"#3#4"())

Body::Matrix{Float64}
    @ REPL[2]:1 within `#3`
1 ─ %1 = Main.cos::Core.Const(cos)
│   %2 = (1:4)::Core.Const(1:4)
│   %3 = Main.Diagonal(%2)::Core.Const([1 0 0 0; 0 2 0 0; 0 0 3 0; 0 0 0 4])
│   %4 = Base.broadcasted(%1, %3)::Core.Const(Base.Broadcast.Broadcasted(cos, ([1 0 0 0; 0 2 0 0; 0 0 3 0; 0 0 0 4],)))
│   %5 = Base.materialize(%4)::Matrix{Float64}
└──      return %5
Select a call to descend into or ↩ to ascend. [q]uit. [b]ookmark.
Toggles: [w]arn, [h]ide type-stable statements, [t]ype annotations, [s]yntax highlight for Source/LLVM/Native.
Show: [S]ource code, [A]ST, [T]yped code, [L]LVM IR, [N]ative code
Actions: [E]dit source code, [R]evise and redisplay
   %2 = < concrete eval > Colon(::Core.Const(1),::Core.Const(4))::Core.Const(1:4)
   %3 = < concrete eval > Diagonal(::Core.Const(1:4))::Core.Const([1 0 0 0; 0 2 0 0; 0 0 3 0; 0 0 0 4])
 • %4 = < concrete eval > broadcasted(::Core.Const(cos),::Core.Const([1 0 0 0; 0 2 0 0; 0 0 3 0; 0 0 0 4]))::…
   %5 = < constprop > materialize(::Core.Const(Base.Broadcast.Broadcasted(cos, ([1 0 0 0; 0 2 0 0; 0 0 3 0; 0 0 0 4],))))::Matrix{Float64}
   ↩
[ Info: tracking Base
broadcasted(f, arg1, args...) @ Base.Broadcast ~/packages/julias/julia-1.9/share/julia/base/broadcast.jl:1308
┌ Warning: Some line information is missing, type-assignment may be incomplete
└ @ Cthulhu ~/Dropbox/JuliaPackages/Cthulhu.jl/src/codeview.jl:117
1308 function broadcasted(f::typeof(cos), arg1::Diagonal{Int64, UnitRange{Int64}}, args::Tuple{}...)::Base.Broadcast.Broadcasted{LinearAlgebra.StructuredMatrixStyle{Diagonal{Int64, UnitRange{Int64}}}, Nothing, typeof(cos), Tuple{Diagonal{Int64, UnitRange{Int64}}}}
1309     arg1′::Diagonal{Int64, UnitRange{Int64}} = broadcastable(arg1::Diagonal{Int64, UnitRange{Int64}})::Diagonal{Int64, UnitRange{Int64}}
1310     args′::Core.Const(()) = map(broadcastable, args::Tuple{})::Tuple{}
1311     broadcasted(combine_styles(arg1′, args′...), f::typeof(cos), arg1′::Diagonal{Int64, UnitRange{Int64}}, args′::Tuple{}...)::Base.Broadcast.Broadcasted{LinearAlgebra.StructuredMatrixStyle{Diagonal{Int64, UnitRange{Int64}}}, Nothing, typeof(cos), Tuple{Diagonal{Int64, UnitRange{Int64}}}}
1312 end
Select a call to descend into or ↩ to ascend. [q]uit. [b]ookmark.
Toggles: [w]arn, [h]ide type-stable statements, [t]ype annotations, [s]yntax highlight for Source/LLVM/Native.
Show: [S]ource code, [A]ST, [T]yped code, [L]LVM IR, [N]ative code
Actions: [E]dit source code, [R]evise and redisplay
   1309 broadcastable(arg1::Diagonal{Int64, UnitRange{Int64}})
   1310 map(broadcastable, args::Tuple{})
   %7 = combine_styles(::Diagonal{Int64, UnitRange{Int64}})::Core.Const(LinearAlgebra.StructuredMatrixStyle{Diagonal{Int64, UnitRange{Int64}}}())
 • 1311 broadcasted(combine_styles(arg1′, args′...), f::typeof(cos), arg1′::Diagonal{Int64, UnitRange{Int64}}, args′::Tuple{}...)
   ↩
broadcasted(::S, f, args...) where S<:BroadcastStyle @ Base.Broadcast ~/packages/julias/julia-1.9/share/julia/base/broadcast.jl:1319
┌ Warning: Some line information is missing, type-assignment may be incomplete
└ @ Cthulhu ~/Dropbox/JuliaPackages/Cthulhu.jl/src/codeview.jl:117
1319  (broadcasted(::(S)::LinearAlgebra.StructuredMatrixStyle{Diagonal{Int64, UnitRange{Int64}}}, f::typeof(cos), args::Tuple{Diagonal{Int64, UnitRange{Int64}}}...) where S<:BroadcastStyle)::Base.Broadcast.Broadcasted{LinearAlgebra.StructuredMatrixStyle{Diagonal{Int64, UnitRange{Int64}}}, Nothing, typeof(cos), Tuple{Diagonal{Int64, UnitRange{Int64}}}} = Broadcasted::Type{Base.Broadcast.Broadcasted}{S::Type{LinearAlgebra.StructuredMatrixStyle{Diagonal{Int64, UnitRange{Int64}}}}}::Type{Base.Broadcast.Broadcasted{LinearAlgebra.StructuredMatrixStyle{Diagonal{Int64, UnitRange{Int64}}}}}(f::typeof(cos), args::Tuple{Diagonal{Int64, UnitRange{Int64}}})
Select a call to descend into or ↩ to ascend. [q]uit. [b]ookmark.
Toggles: [w]arn, [h]ide type-stable statements, [t]ype annotations, [s]yntax highlight for Source/LLVM/Native.
Show: [S]ource code, [A]ST, [T]yped code, [L]LVM IR, [N]ative code
Actions: [E]dit source code, [R]evise and redisplay
 • 1319 Broadcasted::Type{Base.Broadcast.Broadcasted}{S::Type{LinearAlgebra.StructuredMatrixStyle{Diagonal{Int64, UnitRange{Int64}}}}}::Type{Base.Broadcast……………………
   ↩
ERROR: AssertionError: kind(arg) == K"Identifier" || is_operator(arg)
Stacktrace:
  [1] (::TypedSyntax.var"#argidentifier#10")(arg::JuliaSyntax.TreeNode{TypedSyntax.TypedSyntaxData})
    @ TypedSyntax ~/.julia/packages/TypedSyntax/uinhI/src/node.jl:119
  [2] map_signature!(sig::JuliaSyntax.TreeNode{TypedSyntax.TypedSyntaxData}, src::Core.CodeInfo)
    @ TypedSyntax ~/.julia/packages/TypedSyntax/uinhI/src/node.jl:143
  [3] JuliaSyntax.TreeNode{TypedSyntax.TypedSyntaxData}(rootnode::JuliaSyntax.TreeNode{JuliaSyntax.SyntaxData}, src::Core.CodeInfo, mappings::Vector{Vector{Union{JuliaSyntax.TreeNode{JuliaSyntax.SyntaxData}, JuliaSyntax.TreeNode{TypedSyntax.TypedSyntaxData}}}}, symtyps::IdDict{JuliaSyntax.TreeNode{JuliaSyntax.SyntaxData}, Any})
    @ TypedSyntax ~/.julia/packages/TypedSyntax/uinhI/src/node.jl:71
  [4] tsn_and_mappings(m::Method, src::Core.CodeInfo, rt::Any; warn::Bool, strip_macros::Bool, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ TypedSyntax ~/.julia/packages/TypedSyntax/uinhI/src/node.jl:45
  [5] tsn_and_mappings
    @ ~/.julia/packages/TypedSyntax/uinhI/src/node.jl:28 [inlined]
  [6] #get_typed_sourcetext#32
    @ ~/Dropbox/JuliaPackages/Cthulhu.jl/src/reflection.jl:350 [inlined]
  [7] get_typed_sourcetext(mi::Core.MethodInstance, ::Core.Compiler.IRCode, rt::Any; kwargs::Base.Pairs{Symbol, Bool, Tuple{Symbol}, NamedTuple{(:warn,), Tuple{Bool}}})
    @ Cthulhu ~/Dropbox/JuliaPackages/Cthulhu.jl/src/reflection.jl:357
  [8] find_callsites(interp::Cthulhu.CthulhuInterpreter, CI::Core.Compiler.IRCode, stmt_infos::Vector{Core.Compiler.CallInfo}, mi::Core.MethodInstance, slottypes::Vector{Any}, optimize::Bool, annotate_source::Bool)
    @ Cthulhu ~/Dropbox/JuliaPackages/Cthulhu.jl/src/reflection.jl:28
  [9] _descend(term::REPL.Terminals.TTYTerminal, interp::Cthulhu.CthulhuInterpreter, curs::Cthulhu.CthulhuCursor; override::Cthulhu.SemiConcreteCallInfo, debuginfo::Cthulhu.DInfo.DebugInfo, optimize::Bool, interruptexc::Bool, iswarn::Bool, hide_type_stable::Bool, verbose::Nothing, remarks::Bool, with_effects::Bool, inline_cost::Bool, type_annotations::Bool, annotate_source::Bool)
    @ Cthulhu ~/Dropbox/JuliaPackages/Cthulhu.jl/src/Cthulhu.jl:496
 [10] _descend(term::REPL.Terminals.TTYTerminal, interp::Cthulhu.CthulhuInterpreter, curs::Cthulhu.CthulhuCursor; override::Nothing, debuginfo::Cthulhu.DInfo.DebugInfo, optimize::Bool, interruptexc::Bool, iswarn::Bool, hide_type_stable::Bool, verbose::Nothing, remarks::Bool, with_effects::Bool, inline_cost::Bool, type_annotations::Bool, annotate_source::Bool) (repeats 2 times)
    @ Cthulhu ~/Dropbox/JuliaPackages/Cthulhu.jl/src/Cthulhu.jl:629
 [11] _descend(term::REPL.Terminals.TTYTerminal, interp::Cthulhu.CthulhuInterpreter, curs::Cthulhu.CthulhuCursor; override::Nothing, debuginfo::Symbol, optimize::Bool, interruptexc::Bool, iswarn::Bool, hide_type_stable::Bool, verbose::Nothing, remarks::Bool, with_effects::Bool, inline_cost::Bool, type_annotations::Bool, annotate_source::Bool)
    @ Cthulhu ~/Dropbox/JuliaPackages/Cthulhu.jl/src/Cthulhu.jl:629
 [12] _descend
    @ ~/Dropbox/JuliaPackages/Cthulhu.jl/src/Cthulhu.jl:413 [inlined]
 [13] #_descend#107
    @ ~/Dropbox/JuliaPackages/Cthulhu.jl/src/Cthulhu.jl:765 [inlined]
 [14] _descend
    @ ~/Dropbox/JuliaPackages/Cthulhu.jl/src/Cthulhu.jl:765 [inlined]
 [15] #_descend#110
    @ ~/Dropbox/JuliaPackages/Cthulhu.jl/src/Cthulhu.jl:781 [inlined]
 [16] __descend_with_error_handling(args::Any; terminal::Any, kwargs::Base.Pairs{Symbol, V, Tuple{Vararg{Symbol, N}}, NamedTuple{names, T}} where {V, N, names, T<:Tuple{Vararg{Any, N}}})
    @ Cthulhu ~/Dropbox/JuliaPackages/Cthulhu.jl/src/Cthulhu.jl:239
 [17] _descend_with_error_handling(f::Any, argtypes::Any; kwargs::Base.Pairs{Symbol, Bool, Tuple{Symbol}, NamedTuple{(:iswarn,), Tuple{Bool}}})
    @ Cthulhu ~/Dropbox/JuliaPackages/Cthulhu.jl/src/Cthulhu.jl:228
 [18] descend_code_warntype(::Any, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Cthulhu ~/Dropbox/JuliaPackages/Cthulhu.jl/src/Cthulhu.jl:204
 [19] descend_code_warntype(::Any, ::Any)
    @ Cthulhu ~/Dropbox/JuliaPackages/Cthulhu.jl/src/Cthulhu.jl:204
 [20] top-level scope
    @ REPL[2]:1
```

--- a/TypedSyntax/src/node.jl
+++ b/TypedSyntax/src/node.jl
@@ -116,6 +116,9 @@ function map_signature!(sig::TypedSyntaxNode, src::CodeInfo)
             @assert kind(arg) == K"quote"
             arg = only(children(arg))
         end
+        if kind(arg) == K"curly"
+            arg = first(children(arg))
+        end
         @assert kind(arg) == K"Identifier" || is_operator(arg)
         return arg, defaultval
     end

--- a/TypedSyntax/test/runtests.jl
+++ b/TypedSyntax/test/runtests.jl
@@ -426,6 +426,12 @@ include("test_module.jl")
     sig, body = children(tsn)
     node = child(body, 1)
     @test node.typ === Type{Float64}
+    tsn = TypedSyntaxNode(TSN.DefaultArray{Float32}, (Vector{Int}, Int))
+    sig, body = children(tsn)
+    @test kind(sig) == K"where"
+    @test kind(child(sig, 1)) == K"call"
+    f = child(sig, 1, 1)
+    @test kind(f) == K"curly" && f.typ === Type{TSN.DefaultArray{Float32}}
 
     # Field access & a more complex example
     tsn = TypedSyntaxNode(Base.getindex, (TSN.DefaultArray{Float64,2,Matrix{Float64}}, Int, Int))

--- a/TypedSyntax/test/test_module.jl
+++ b/TypedSyntax/test/test_module.jl
@@ -80,7 +80,14 @@ struct DefaultArray{T,N,A<:AbstractArray{T,N}} <: AbstractArray{T,N}
     parentarray::A
     defaultvalue::T
 end
-DefaultArray(parentarray, defaultvalue) = DefaultArray{ndims(parentarray)}(parentarray, defaultvalue)
+function DefaultArray{T}(parentarray::AbstractArray, defaultvalue) where T
+    p = convert(AbstractArray{T}, parentarray)
+    return DefaultArray{T, ndims(parentarray), typeof(p)}(p, convert(T, defaultvalue))
+end
+function DefaultArray(parentarray::AbstractArray, defaultvalue)
+    T = promote_type(eltype(parentarray), typeof(defaultvalue))
+    return DefaultArray{T}(parentarray, defaultvalue)
+end
 Base.getindex(a::DefaultArray{T,N}, i::Vararg{Int,N}) where {T,N} = checkbounds(Bool, a, i...) ? a.parentarray[i...] : a.defaultvalue
 Base.size(a::DefaultArray) = size(a.parentarray)
 

